alacritty/src/renderer/solidrect.rs

```diff
@@ -0,0 +1,200 @@
+                0,
+                2,
+                gl::FLOAT,
+                gl::FALSE,
+                (size_of::<Vertex>()) as _,
+                ptr::null(),
+            );
+            gl::EnableVertexAttribArray(0);
+
+            // Color
+            gl::VertexAttribPointer(
+                1,
+                4,
+                gl::UNSIGNED_BYTE,
+                gl::TRUE,
+                (size_of::<Vertex>()) as _,
+                offset_of!(Vertex, color) as *const _,
+            );
+            gl::EnableVertexAttribArray(1);
+
+            gl::BindVertexArray(0);
+            gl::BindBuffer(gl::ARRAY_BUFFER, 0);
+        }
+
+        Self { vao, vbo, ebo, indices: Vec::new(), vertices: Vec::new(), uploaded_indices: 0 }
+    }
+
+    pub fn draw(
+        &mut self,
+        program: &RectShaderProgram,
+        size_info: &SizeInfo,
+        rects: Vec<RenderRect>,
+    ) {
+        if rects.is_empty() {
+            return;
+        }
+
+        // Prepare common state
+        unsafe {
+            // Remove padding from viewport.
+            gl::Viewport(0, 0, size_info.width() as i32, size_info.height() as i32);
+
+            gl::Enable(gl::BLEND);
+            gl::BlendFuncSeparate(gl::SRC_ALPHA, gl::ONE_MINUS_SRC_ALPHA, gl::SRC_ALPHA, gl::ONE);
+
+            // Setup bindings. VAO will set up attribs and EBO, but not VBO.
+            gl::BindVertexArray(self.vao);
+            gl::BindBuffer(gl::ARRAY_BUFFER, self.vbo);
+
+            gl::UseProgram(program.id);
+        }
+
+        let center_x = size_info.width() / 2.;
+        let center_y = size_info.height() / 2.;
+
+        for rect in &rects {
+            self.append_rect(center_x, center_y, rect);
+        }
+
+        self.draw_accumulated();
+    }
+
+    fn append_rect(&mut self, center_x: f32, center_y: f32, rect: &RenderRect) {
+        if rect.alpha <= 0. {
+            return;
+        }
+
+        assert!(self.vertices.len() <= MAX_U16_INDICES - 4);
+
+        // Calculate rectangle position.
+        let x = (rect.x - center_x) / center_x;
+        let y = -(rect.y - center_y) / center_y;
+        let width = rect.width / center_x;
+        let height = rect.height / center_y;
+        let color = Rgba {
+            r: rect.color.r,
+            g: rect.color.g,
+            b: rect.color.b,
+            a: (rect.alpha * 255.) as u8,
+        };
+
+        self.vertices.extend_from_slice(&[
+            Vertex { x, y, color },
+            Vertex { x, y: y - height, color },
+            Vertex { x: x + width, y, color },
+            Vertex { x: x + width, y: y - height, color },
+        ]);
+
+        if self.vertices.len() == MAX_U16_INDICES {
+            self.draw_accumulated();
+        }
+    }
+
+    fn draw_accumulated(&mut self) {
+        if self.vertices.is_empty() {
+            return;
+        }
+
+        // Generate new indices in index buffer on-demand
+        assert!(self.indices.len() % 6 == 0);
```

Comment:
Wait, how do `indices` work. You store this globally on the rect renderer and they're initialized once when the renderer is created (which happens once at launch iirc).

Then you push to indices whenever an accumulated draw is performed.

~~Isn't this leaking memory? Where are the indices cleared up? Shouldn't they be re-created every time `draw_accumulated` is called?~~

~~If the goal is to prevent reallocation every time and keep a bit of memory around for the indices, one could call `clear` on the vec before drawing, but I feel like it would probably be fine to just recreate the Vec?~~

**Edit:** You do indeed seem to be calling `clear` at least. So this isn't a memory leak. But I'm still not sure if it isn't much easier to just re-create the vec in `draw_accumulated` and would be curious what the performance impact of that is.

Also as a sidenote: Not a massive fan from shipping assertions in production code. Usually those are just pointless branches that should never be hit. Generally we only have those enabled in debug mode.

Reply:
I think I should just pre-allocate and pre-upload indices upfront. That would be the simplest way :).

Asserts are intended for debug mode only and are there to provide visibility into invariants expected. I can certainly remove them if you feel that they provide no value.

Reply:
>Asserts are intended for debug mode only and are there to provide visibility into invariants expected. I can certainly remove them if you feel that they provide no value.

if you want them for debug, use `debug_assert`.

Reply:
Replaced a single remaining assert with debug_assert